api/transport/stream.go

```diff
@@ -0,0 +1,165 @@
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package transport
+
+import (
+	"context"
+	"io"
+
+	"go.uber.org/yarpc/yarpcerrors"
+)
+
+// StreamRequest is used to establish a streaming request.  It contains basic
+// Metadata information, and any Stream-specfic configuration for creating a
+// stream.
+type StreamRequest struct {
+	Meta *RequestMeta
+}
+
+type serverStreamOpts struct{}
+
+// ServerStreamOption are options to configure a ServerStream
+type ServerStreamOption interface {
+	apply(*serverStreamOpts)
+}
+
+// NewServerStream will create a new ServerStream.
+func NewServerStream(s Stream, options ...ServerStreamOption) (*ServerStream, error) {
+	if s == nil {
+		return nil, yarpcerrors.InvalidArgumentErrorf("non-nil stream is required")
+	}
+	return &ServerStream{stream: s}, nil
+}
+
+// ServerStream represents the Server API of interacting with a Stream.
+type ServerStream struct {
+	stream Stream
+}
+
+// Context returns the context for the stream.
+func (s *ServerStream) Context() context.Context {
+	return s.stream.Context()
+}
+
+// Request contains all the metadata about the request.
+func (s *ServerStream) Request() *StreamRequest {
+	return s.stream.Request()
+}
+
+// SendMessage sends a request over the stream. It blocks until the message
+// has been sent.  In certain implementations, the timeout on the context
+// will be used to timeout the request.
+func (s *ServerStream) SendMessage(ctx context.Context, msg *StreamMessage) error {
+	return s.stream.SendMessage(ctx, msg)
+}
+
+// ReceiveMessage blocks until a message is received from the connection. It
+// returns an io.Reader with the contents of the message.
+func (s *ServerStream) ReceiveMessage(ctx context.Context) (*StreamMessage, error) {
+	return s.stream.ReceiveMessage(ctx)
+}
+
+type clientStreamOpts struct{}
+
+// ClientStreamOption is an option for configuring a client stream.
+type ClientStreamOption interface {
+	apply(*clientStreamOpts)
```

Comment:
optional: since clientStreamOpts isn't being used at this time, you can change this interface to

```
type ClientStreamOption interface {
  unimplemented()
}
```

This can be changed to `apply(*clientStreamOpts)` with the corresponding
struct being declared when the need arises.

We do something similar in Fx.